### ISSUE TITLE
Execution engine - Handle map and reduce

### ISSUE THREAD

**Issue by anibalsolon** [radiome-lab/radiome#30]

**Comment by anibalsolon** [comment]
@ccraddock @puorc Please, whenever you have time, take a look and let's discuss!

**Comment by puorc** [comment]
Hi Anibal, the map operator is nice! But I feel a little confused on 'reduce the entries from job_c.out_field over the dimension job_a.list_field to job_c.field'. Could you elaborate a bit more on ' dimension'? Does it represent one attribute of the return dict?

**Comment by anibalsolon** [comment]
Yes, so every time we map a list of N items to N nodes, we create a dimension for a node: from a dot (a simple workflow, in a nildimensional space), we go to a line with N points.

![image](https://user-images.githubusercontent.com/562525/77859801-0a1d0900-71d1-11ea-8800-316633f166d1.png)

Mapping a list:

![image](https://user-images.githubusercontent.com/562525/77859806-13a67100-71d1-11ea-8406-62aafe10e9c2.png)

Mapping two lists, so the input of each node is the combination of each pair of items in both lists:

![image](https://user-images.githubusercontent.com/562525/77859818-1bfeac00-71d1-11ea-8911-d5c496537560.png)

And, as we reduce each of the dimensions, we create lists of the results of the nodes. It is the reducing operation, but instead of summarizing the results (like in an average), we are creating lists.

![image](https://user-images.githubusercontent.com/562525/77859830-2a4cc800-71d1-11ea-93e4-efbc9fd8b7c8.png)

If we reduced/squashed both dimensions at the same time, and inputted it to a node, the value would be a list of lists (i.e. a matrix) of the results.

**Comment by puorc** [comment]
Thank you for the illustration. It's nice to work on.
